api/src/main/java/org/apache/iceberg/stats/StatsUtil.java

```diff
@@ -0,0 +1,223 @@
+  static final int VALUE_COUNT_OFFSET = 0;
+  static final int NULL_VALUE_COUNT_OFFSET = 1;
+  static final int NAN_VALUE_COUNT_OFFSET = 2;
+  static final int AVG_VALUE_SIZE_OFFSET = 3;
+  static final int MAX_VALUE_SIZE_OFFSET = 4;
+  static final int LOWER_BOUND_OFFSET = 5;
+  static final int UPPER_BOUND_OFFSET = 6;
+  static final int NUM_STATS_PER_COLUMN = 200;
+  static final int RESERVED_FIELD_IDS = 200;
+  static final int DATA_SPACE_FIELD_ID_START = 10_000;
+  static final int METADATA_SPACE_FIELD_ID_START = 2_147_000_000;
+  static final int RESERVED_FIELD_IDS_START = Integer.MAX_VALUE - RESERVED_FIELD_IDS;
+  // support stats for only up to this amount of data field IDs
+  static final int MAX_DATA_FIELD_ID = 1_000_000;
+  static final int MAX_DATA_STATS_FIELD_ID = 200_010_000;
+
+  private StatsUtil() {}
+
+  public static int statsFieldIdForField(int fieldId) {
+    return fieldId >= RESERVED_FIELD_IDS_START
+        ? statsFieldIdForReservedField(fieldId)
+        : statsFieldIdForDataField(fieldId);
+  }
+
+  private static int statsFieldIdForDataField(int fieldId) {
+    long statsFieldId = DATA_SPACE_FIELD_ID_START + NUM_STATS_PER_COLUMN * (long) fieldId;
+    if (fieldId < 0 || fieldId > MAX_DATA_FIELD_ID || statsFieldId < 0) {
+      return -1;
+    }
+
+    return (int) statsFieldId;
+  }
+
+  private static int statsFieldIdForReservedField(int fieldId) {
+    int offset = RESERVED_FIELD_IDS - (Integer.MAX_VALUE - fieldId);
+
+    long statsFieldId = METADATA_SPACE_FIELD_ID_START + NUM_STATS_PER_COLUMN * (long) offset;
+    if (statsFieldId < 0 || statsFieldId > RESERVED_FIELD_IDS_START) {
+      // ID overflows
+      return -1;
+    }
+
+    return (int) statsFieldId;
+  }
+
+  public static int fieldIdForStatsField(int statsFieldId) {
+    if (statsFieldId < DATA_SPACE_FIELD_ID_START || statsFieldId % NUM_STATS_PER_COLUMN != 0) {
+      return -1;
+    }
+
+    return statsFieldId < METADATA_SPACE_FIELD_ID_START
+        ? fieldIdForStatsFieldFromDataField(statsFieldId)
+        : fieldIdForStatsFieldFromReservedField(statsFieldId);
+  }
+
+  private static int fieldIdForStatsFieldFromDataField(int statsFieldId) {
+    return Math.max(-1, (statsFieldId - DATA_SPACE_FIELD_ID_START) / NUM_STATS_PER_COLUMN);
+  }
+
+  private static int fieldIdForStatsFieldFromReservedField(int statsFieldId) {
+    return Math.max(
+        -1,
+        statsFieldId
+            - RESERVED_FIELD_IDS
+            + (Integer.MAX_VALUE - statsFieldId)
+            + (statsFieldId - METADATA_SPACE_FIELD_ID_START) / NUM_STATS_PER_COLUMN);
+  }
+
+  public static Types.NestedField contentStatsFor(Schema schema) {
+    ContentStatsSchemaVisitor visitor = new ContentStatsSchemaVisitor();
+    Types.NestedField result = TypeUtil.visit(schema, visitor);
+    if (!visitor.skippedFieldIds.isEmpty()) {
+      LOG.warn(
+          "Could not create stats schema for field ids {} of schema: {}",
```

Comment:
should we explain the reason for skipping? stats are only allowed for field id up to 1M.

Reply:
> Invalid field id (like <0) should really fail, as it is an internal error.

I believe we need to be lenient with invalid field IDs as we don't want to fail manifest reading/writing. See also my other [comment](https://github.com/apache/iceberg/pull/13933#discussion_r2413743360) on this topic.

> I am also wondering if we need to print the schema. The schema can be very large (>1M fields) if we really ran into the condition of skipping

Yes you're right, schema could become quite large, so maybe it's better to remove it from the msg